### PR TITLE
Performance and memory improvements to batched torch.linalg.solve

### LINKED ISSUE #71406: torch.linalg.solve updates are causing GPyTorch tests to fail
Since https://github.com/pytorch/pytorch/pull/69752 the tests `test_minres_mat`:

https://github.com/cornellius-gp/gpytorch/blob/2489afb17d53166ee693ebd86b6cf9e152013ce6/test/utils/test_minres.py#L55

and `test_minres_mat_multiple_shifts`:

https://github.com/cornellius-gp/gpytorch/blob/2489afb17d53166ee693ebd86b6cf9e152013ce6/test/utils/test_minres.py#L61

are failing in GPyTorch.

cc @ezyang @gchanan @zou3519 @bdhirsh @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @Balandat

### LINKED ISSUE #71610: `torch.linalg.solve` INTERNAL ASSERT FAILED
### 🐛 Describe the bug

`torch.linalg.solve` can trigger INTERNAL ASSERT FAILED randomly. For example,

```python
import torch
input = torch.rand([2, 0, 5, 5], dtype=torch.complex128)
A = torch.rand([1, 5, 5], dtype=torch.complex128)
torch.linalg.solve(A, input)
# RuntimeError: falseINTERNAL ASSERT FAILED at "../aten/src/ATen/native/LinearAlgebraUtils.h":244, please report a bug to PyTorch. linalg_solve: (Batch element 0): Argument 335551511 has illegal value. Most certainly there is a bug in the implementation calling the backend library.
```
You should run the code a few times to trigger the internal fail.

### Versions

pytorch: 1.10.1

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Previously for single input matrix A and batched matrix B, matrix A was expanded and cloned before computing the LU decomposition and solving the linear system.

With this PR the LU decomposition is computed once for a single matrix and then expanded&cloned if required by a backend library call for the linear system solving.

Here's a basic comparison:
```python
# BEFORE THE PR
In [1]: import torch
In [2]: a = torch.randn(256, 256)
In [3]: b = torch.randn(1024, 256, 2)
In [4]: %%timeit
   ...: torch.linalg.solve(a, b)
   ...:
   ...:
329 ms ± 17.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# WITH THIS PR
In [1]: import torch
In [2]: a = torch.randn(256, 256)
In [3]: b = torch.randn(1024, 256, 2)
In [4]: %%timeit
   ...: torch.linalg.solve(a, b)
   ...:
   ...:
21.4 ms ± 23 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Fixes https://github.com/pytorch/pytorch/issues/71406, fixes https://github.com/pytorch/pytorch/issues/71610

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -2972,9 +2974,36 @@ static void lu_solve_looped_magma(const Tensor& b, const Tensor& lu, const Tenso
   });
 }
 
+namespace {
+
+c10::MaybeOwned<Tensor> maybe_expand_lu(const Tensor& b, const Tensor& lu) {
+  if (batchCount(b) != batchCount(lu)) {
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2980]: This works whenever `b` is a matrix, but not when `B` is a batch of vectors, right?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -2972,9 +2974,36 @@ static void lu_solve_looped_magma(const Tensor& b, const Tensor& lu, const Tenso
...
+    expand_size.insert(expand_size.end(), {lu.size(-2), lu.size(-1)});
+    return c10::MaybeOwned<Tensor>::owned(
+        cloneBatchedColumnMajor(lu.expand(expand_size)));
+  } else {
+    return c10::MaybeOwned<Tensor>::borrowed(lu);
+  }
+}
+
+c10::MaybeOwned<Tensor> maybe_expand_pivots(const Tensor& b,const Tensor& pivots) {
+  if (batchCount(b) != batchCount(pivots.unsqueeze(-1))) {
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2992]: Same as above.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -2861,8 +2861,8 @@ static void apply_lu_solve_looped_magma(const Tensor& b, const Tensor& lu, const
   auto pivots_data = pivots_cpu.data_ptr<magma_int_t>();
 
   auto b_stride = matrixStride(b);
-  auto lu_stride = matrixStride(lu);
-  auto pivots_stride = pivots_cpu.size(-1);
+  auto lu_stride = lu.dim() > 2 ? lu.stride(-3) : 0;
+  auto pivots_stride = pivots_cpu.dim() > 1 ? pivots_cpu.stride(-2) : 0;
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2865]: No need to instantiate a large matrix here! This is great.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -2906,6 +2906,8 @@ static void apply_lu_solve_batched_magma(const Tensor& b, const Tensor& lu, cons
       "Calling torch.lu_solve on a CUDA tensor requires compiling ",
       "PyTorch with MAGMA. Please rebuild with MAGMA.");
 #else
+  TORCH_INTERNAL_ASSERT(batchCount(b) == batchCount(lu), "batch_size of b and lu must be the same");
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2909]: Again, same issue with vectors, right?

**Review by lezcano — COMMENTED** [review]
Great optimisation when you have a single matrix as the lhs and a batch of matrices / vectors as the rhs.
Now, I believe you forgot to handle the pesky corner case of solving `Ax = b` for `b` a vector or a batch of vectors (I don't really like the fact that we even support this one, but well)

**Comment by IvanYashchuk** [comment]
When `B` is a batch of vectors then the rule is that batches of `A` and `B` should match. It's not possible to have a single matrix input A and a batch of vectors.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2980]: When we reach this level of code `B` is always a matrix or batch of matrices.

**Review by lezcano — APPROVED** [review]
You are totally right. This is a very nice optimisation.

**Review by mruberry — APPROVED** [review]
Stamped!

Awesome perf improvement, @IvanYashchuk. We should wait to CI tests, and they may need to be rerun

**Comment by mruberry** [comment]
@IvanYashchuk would you just rerun the tests on this and ping me? Our CI has recovered from the outages affecting it last week and earlier this week.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -966,7 +965,7 @@ Tensor& linalg_solve_out(const Tensor& input, const Tensor& other, Tensor& resul
   // Now check LAPACK/MAGMA error codes
   // batchCheckErrors(Tensor, char*) calls 'infos = infos.to(kCPU)'
   bool vector_case = linalg_solve_is_vector_rhs(input, other);
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:967]: Internal builds correctly complain this variable is now unused. Is that intentional?

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:967]: Yes, it's intentional. I should remove it.

**Comment by IvanYashchuk** [comment]
@mruberry, I removed the unused variable. I hope internal builds will pass now.

**Comment by IvanYashchuk** [comment]
@mruberry, could you please try importing this pull request?

**Comment by mruberry** [comment]
> @mruberry, could you please try importing this pull request?

Thanks for the ping, Ivan!

**Comment by IvanYashchuk** [comment]
This PR is being reverted with https://github.com/pytorch/pytorch/pull/71421

**Comment by IvanYashchuk** [comment]
This PR has a bug that broadcasting is not handled correctly leading to invalid memory write/reads.

**Comment by mruberry** [comment]
> @Lezcano, @mruberry could you please take another look at this PR? I fixed the problem with broadcasting in [a316fac](https://github.com/pytorch/pytorch/commit/a316face79006635890cb6de66290aed97fc1d30). The same speedup as in the PR description holds. The main problem was that we didn't have tests that check the broadcasting behavior of `torch.linalg.solve`. I also verified that failing tests reported in #71406 pass now.

Would you rebase this and ping me? It has a merge conflict.

**Comment by IvanYashchuk** [comment]
@mruberry, I resolved the merge conflict.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -984,13 +984,8 @@ Tensor& linalg_solve_out(const Tensor& input, const Tensor& other, Tensor& resul
...
   // Now check LAPACK/MAGMA error codes
-  // batchCheckErrors(Tensor, char*) calls 'infos = infos.to(kCPU)'
-  if (input.dim() > 2) {
-    batchCheckErrors(infos, "linalg.solve");
-  } else {
-    singleCheckErrors(infos.item().toInt(), "linalg.solve");
-  }
-
+  // _linalg_check_errors calls 'infos = infos.to(kCPU)'
+  at::_linalg_check_errors(infos, "linalg.solve", input.dim() > 2);
```
Comment (Lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:988]: Nice!

**Review by mruberry — APPROVED** [review]
Thanks @IvanYashchuk!

**Comment by mruberry** [comment]
Unfortunately I didn't realize that our internal system won't let us land the same PR twice. @IvanYashchuk would you open a new PR with this content and just ping me to merge it immediately?

**Comment by IvanYashchuk** [comment]
Interesting, I'm sure we did this before. Anyways I opened a new PR.

**Pull request closed without merging**